src/python/pants/backend/jvm/tasks/jvm_compile/zinc/zinc_compile.py

```diff
@@ -538,6 +545,38 @@ def compile(
             },
         )()
 
+    def always_do_after_compile(self, ctx):
+        self._pass_diagnostics_to_buildstats(ctx)
+
+    def _aggregate_diagnostics(self, lsp_data):
+        # Note: this is not arbitrary. It is exactly every value that the LSP's DiagnosticSeverity allows.
+        # See https://microsoft.github.io/language-server-protocol/specification#diagnostic
+        counts = {
+            "Error": 0,
+            "Warning": 0,
+            "Information": 0,
+            "Hint": 0,
+        }
+        for published_diagnostics in lsp_data:
+            for diagnostic in published_diagnostics["diagnostics"]:
+                severity = diagnostic["severity"]
+                counts[severity] += 1
+        return counts
+
+    def _pass_diagnostics_to_buildstats(self, ctx):
+        diagnostics_file = self._diagnostics_out(ctx)
+        if not os.path.exists(diagnostics_file):
+            return
+        with open(diagnostics_file) as json_diagnostics:
+            data = json.load(json_diagnostics)
+            counts = self._aggregate_diagnostics(data)
+            self.context.log.debug(f"Reporting number of diagnostics for: {ctx.target}")
+            for (severity, count) in counts.items():
+                self.context.log.debug(f"    {severity}: {count}")
+                self.context.run_tracker.report_target_info(
+                    self.options_scope, ctx.target, ["n_diagnostics", severity], count
```

Comment:
Is the idea behind the `n_` prefix that it's the count, and that it reserves `diagnostics` for future use of actually including more structured information about the diagnostics? If so, I'd maybe call it `diagnostic_counts`? If not, I'm curious as to what it's about :)

Reply:
Yes, that's what it means. Happy to rename to `diagnostic_counts` 🙂